ksqldb-execution/src/main/java/io/confluent/ksql/execution/plan/ExecutionKeyFactory.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright 2019 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License; you may not use this file
+ * except in compliance with the License.  You may obtain a copy of the License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.execution.plan;
+
+import io.confluent.ksql.execution.builder.KsqlQueryBuilder;
+import io.confluent.ksql.execution.context.QueryContext;
+import io.confluent.ksql.schema.ksql.PhysicalSchema;
+import io.confluent.ksql.serde.FormatInfo;
+import io.confluent.ksql.serde.WindowInfo;
+import io.confluent.ksql.testing.EffectivelyImmutable;
+import java.util.Objects;
+import org.apache.kafka.common.serialization.Serde;
+import org.apache.kafka.connect.data.Struct;
+import org.apache.kafka.streams.kstream.Windowed;
+
+/**
+ * The {@code ExecutionKeyFactory} is in charge of creating the keys
+ * and the Serdes for the keys of any particular execution step.
+ *
+ * @param <K> the type of the key, usually either {@code Struct}
+ *            or {@code Windowed<Struct>}
+ */
+@EffectivelyImmutable
+public interface ExecutionKeyFactory<K> {
+
+  Serde<K> buildKeySerde(
+      FormatInfo format,
+      PhysicalSchema physicalSchema,
+      QueryContext queryContext
+  );
+
+  /**
+   * @return a new {@code ExecutionKeyFactory}
+   */
+  ExecutionKeyFactory<K> withQueryBuilder(KsqlQueryBuilder builder);
+
+  /**
+   * @return a new key of type {@code K} given the struct conents of the key
```

Comment:
```suggestion
   * @return a new key of type {@code K} given the struct contents of the new key and the old key being updated
```
Would it make more sense to rename this to something like `updateKey`? I was surprised to see the old key required in a method called `constructKey`, but maybe `updateKey` is even more confusing. `getNewKey`? `constructNewKey`?